janus.c

```diff
@@ -2930,233 +2795,28 @@ json_t *janus_plugin_handle_sdp(janus_plugin_session *plugin_session, janus_plug
-					janus_ice_trickle *trickle = (janus_ice_trickle *)temp->data;
-					g_list_free(temp);
-					if(trickle == NULL)
-						continue;
-					if((janus_get_monotonic_time() - trickle->received) > 45*G_USEC_PER_SEC) {
-						/* FIXME Candidate is too old, discard it */
-						janus_ice_trickle_destroy(trickle);
-						/* FIXME We should report that */
-						continue;
-					}
-					json_t *candidate = trickle->candidate;
-					if(candidate == NULL) {
-						janus_ice_trickle_destroy(trickle);
-						continue;
-					}
-					if(json_is_object(candidate)) {
-						/* We got a single candidate */
-						int error = 0;
-						const char *error_string = NULL;
-						if((error = janus_ice_trickle_parse(ice_handle, candidate, &error_string)) != 0) {
-							/* FIXME We should report the error parsing the trickle candidate */
-						}
-					} else if(json_is_array(candidate)) {
-						/* We got multiple candidates in an array */
-						JANUS_LOG(LOG_VERB, "[%"SCNu64"] Got multiple candidates (%zu)\n", ice_handle->handle_id, json_array_size(candidate));
-						if(json_array_size(candidate) > 0) {
-							/* Handle remote candidates */
-							size_t i = 0;
-							for(i=0; i<json_array_size(candidate); i++) {
-								json_t *c = json_array_get(candidate, i);
-								/* FIXME We don't care if any trickle fails to parse */
-								janus_ice_trickle_parse(ice_handle, c, NULL);
-							}
-						}
-					}
-					/* Done, free candidate */
-					janus_ice_trickle_destroy(trickle);
-				}
-			}
-			/* This was an answer, check if it's time to start ICE */
-			if(janus_flags_is_set(&ice_handle->webrtc_flags, JANUS_ICE_HANDLE_WEBRTC_TRICKLE) &&
-					!janus_flags_is_set(&ice_handle->webrtc_flags, JANUS_ICE_HANDLE_WEBRTC_ALL_TRICKLES)) {
-				JANUS_LOG(LOG_VERB, "[%"SCNu64"]   -- ICE Trickling is supported by the browser, waiting for remote candidates...\n", ice_handle->handle_id);
-				janus_flags_set(&ice_handle->webrtc_flags, JANUS_ICE_HANDLE_WEBRTC_START);
-			} else {
-				JANUS_LOG(LOG_VERB, "[%"SCNu64"] Done! Sending connectivity checks...\n", ice_handle->handle_id);
-				if(ice_handle->audio_id > 0) {
-					janus_ice_setup_remote_candidates(ice_handle, ice_handle->audio_id, 1);
-					if(!janus_flags_is_set(&ice_handle->webrtc_flags, JANUS_ICE_HANDLE_WEBRTC_RTCPMUX))	/* http://tools.ietf.org/html/rfc5761#section-5.1.3 */
-						janus_ice_setup_remote_candidates(ice_handle, ice_handle->audio_id, 2);
-				}
-				if(ice_handle->video_id > 0) {
-					janus_ice_setup_remote_candidates(ice_handle, ice_handle->video_id, 1);
-					if(!janus_flags_is_set(&ice_handle->webrtc_flags, JANUS_ICE_HANDLE_WEBRTC_RTCPMUX))	/* http://tools.ietf.org/html/rfc5761#section-5.1.3 */
-						janus_ice_setup_remote_candidates(ice_handle, ice_handle->video_id, 2);
-				}
-				if(ice_handle->data_id > 0) {
-					janus_ice_setup_remote_candidates(ice_handle, ice_handle->data_id, 1);
-				}
-			}
+			janus_request_ice_handle_answer(ice_handle, audio, video, data, NULL);
```

Comment:
I noticed this when valgrind complained to me about a null pointer dereference in `janus_request_ice_handle_answer`. This code locks the `ice_handle->mutex`, presumably to avoid racing against ICE handle destruction. However, `janus_ice_handle_destroy` (called above when a session is destroyed), which I believe could free the various stuff required by `janus_request_ice_handle_answer`, apparently can be called without the ICE handle's mutex being locked; instead, the session's mutex is locked. Should these two agree on which mutex they are using?

Reply:
Related: I got another valgrind result where the mutex itself, as well as the handle, was freed before we tried to lock on it in this block. [See here.](https://gist.github.com/mquander/7c935aad1a831bff21d9e1ae7c92e7c0)

Reply:
`janus_ice_handle_destroy` doesn't destroy anything by itself, it only unrefs and forces threads to quit. WebRTC resources are destroyed by `janus_ice_webrtc_free` which does lock the same mutex you referenced. You should tell us what's calling `janus_request_ice_handle_answer`: a remote request (remote answer) or your plugin (local answer)? Is this being interrupted by a detach?

As to you pushing an event after `destroy_session`, this should be handled already, as in the core we keep track of valid plugin sessions (see `janus_plugin_session_is_alive`) and if we get one that isn't we dump it. Timing and a race condition may be what's happening here, though, so if you can provide more info on what's going on it will help.

As a side note, don't mix valgrind and libasan, as I believe the two step on one another due to how they keep track of memory allocations their own way. I often use libasan as a lighter alternative to valgrind (valgrind does slow down Janus A LOT, which can make some race conditions easier to trigger but also sometimes cause weird behaviours), although valgrind can give you core dumps you can inspect with gdb.

Reply:
Something we should probably do is increase references when we get in `janus_plugin_push_event` and similar methods, as otherwise the references to the handle and other structs can indeed get to zero while we're using them elsewhere. After all, that's the whole point of using references in the first place...

Reply:
One more note while I'm looking at your gist dump. Right now a core handles is always freed *before* the plugin handle: in fact, a `janus_refcount_decrease` for the plugin handle is in `janus_ice_free`, which is called when the core handle's ref count gets to zero. I guess we did that to allow more flexibility in plugins, as it leaves them with a valid plugin handle pointer (that they can't really use, of course), e.g., to access its opaque properties, even if the core freed everything else.

I'm wondering if this shouldn't be the other way around, now: that is, as long as the plugin handle is alive, the handle in the core still exists as a struct (not a valid handle you can use via API). As soon as the plugin handle refs get to zero, the handle is unreffed to. This would also mean plugins should be much more careful when accessing the plugin handle, though: considering typically a plugin handle only has two refs (one in the core, one in the plugin when `create_session` is called) it means that an unref in the plugin (e.g., in `destroy_session`) would invalidate that pointer much sooner: this can cause problems if the plugin is multithreaded, as they often are, as there need to be additional checks all over the place to be sure the plugin is not accessing a plugin session that's been deallocated. Of course, the plugin can also decide to be smarter with respect to this, and not unref in `destroy_session` but according to different policies and its internal logic (e.g., notifying internal threads about the plugin session being gone), so that it can only unref when it's really sure it won't be needed anymore.

Just food for thought for now, feedback welcome!

Reply:
> You should tell us what's calling janus_request_ice_handle_answer: a remote request (remote answer) or your plugin (local answer)? Is this being interrupted by a detach?

My plugin is calling it with an answer it created.

I haven't seen this again yet since I stopped calling `push_event` after `destroy_session`, so it's not as high a priority for me right now as the other crash in this thread. I'll let you know if I dig into it more.

Reply:
Still a problem we need to solve, though... I'll see if I can replicate the same behaviour locally. If you have feedback as a plugin writer on my last post that would help too!

Reply:
For now, added references to `janus_plugin_push_event` and `janus_plugin_close_pc`, which should help with the problem you were experiencing.

Reply:
> If you have feedback as a plugin writer on my last post that would help too!

Sure. My guess is that your idea would make my life easier. As a Rust plugin writer in particular, it's super easy for me to do my own memory management on my session handle. I can define a struct that looks like

```
struct SessionData {
  handle: *janus_plugin_session,
  data: // whatever state i want
}
```
and then write a Rust destructor for that struct that decreases the Janus refcount on `handle` only when the struct is destroyed. Rust's ownership stuff will make sure that there's no chance that I can accidentally use the session data after destruction. So this part:

> This would also mean plugins should be much more careful when accessing the plugin handle, though: considering typically a plugin handle only has two refs (one in the core, one in the plugin when create_session is called) it means that an unref in the plugin (e.g., in destroy_session) would invalidate that pointer much sooner: this can cause problems if the plugin is multithreaded, as they often are, as there need to be additional checks all over the place to be sure the plugin is not accessing a plugin session that's been deallocated.

is actually very easy for me to do correctly. For example, in my current plugin, I put a Rust reference counting wrapper around the session data structs, and then take out a weak reference to them when I put them on the queue for signalling message processing. When the processing thread pops them off the queue, it only continues if the reference is still live.

As such, your idea eliminates a failure mode that I have to carefully keep an eye on (pushing data using an allocated but "expired" plugin handle pointer) and replaces it with a failure mode that the compiler is designed to help me avoid (incorrectly choosing when to free the session state that I allocate.) I guess the calculus might look a lot different for someone writing a plugin in C, though.